File: plansys2_bt_example/src/charge_action_node.cpp
// Copyright 2019 Intelligent Robotics Lab
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <memory>


#include "behavior_tree_nodes/ApproachObject.hpp"
#include "behavior_tree_nodes/OpenGripper.hpp"
#include "behavior_tree_nodes/CloseGripper.hpp"

#include "plansys2_executor/ActionBTExecutorClient.hpp"

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "ament_index_cpp/get_package_share_directory.hpp"

class ChargeAction : public plansys2::ActionBTExecutorClient
{
public:
  ChargeAction()
  : plansys2::ActionBTExecutorClient(
      "charge",
      ament_index_cpp::get_package_share_directory("plansys2_bt_example") +
      "/behavior_trees_xml/charge.xml")
  {
    factory_.registerNodeType<plansys2_bt_example::ApproachObject>("ApproachObject");
    factory_.registerNodeType<plansys2_bt_example::OpenGripper>("OpenGripper");
    factory_.registerNodeType<plansys2_bt_example::CloseGripper>("CloseGripper");
  }
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<ChargeAction>();

  rclcpp::spin(node->get_node_base_interface());

  rclcpp::shutdown();

  return 0;
}

File: plansys2_simple_example_py/plansys2_simple_example_py/charge_action_node.py
#!/usr/bin/python3

# Copyright 2023 Intelligent Robotics Lab
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import rclpy
from rclpy.parameter import Parameter, ParameterType

from plansys2_support_py.ActionExecutorClient import ActionExecutorClient


class ChargeAction(ActionExecutorClient):

    def __init__(self):
        super().__init__('charge', 0.5)
        self.progress_ = 0.0

    def do_work(self):
        if self.progress_ < 1.0:
            self.progress_ += 0.05
            self.send_feedback(self.progress_, 'Charge running')
        else:
            self.finish(True, 1.0, 'Charge completed');
            self.progress_ = 0.0
        
        self.get_logger().info('charging ... {}'.format(self.progress_))

def main(args=None):
    rclpy.init(args=args)

    node = ChargeAction()
    node.set_parameters([Parameter(name='action_name', value='charge')])

    node.trigger_configure()

    rclpy.spin(node)

    rclpy.shutdown()

if __name__ == '__main__':
    main()

